### PR TITLE
Fe post creation

### PR DESCRIPTION
**Pull request by LukaszBielsko**
front end CRUD implemented

### CODE REVIEW

File: frontend/src/store/actionTypes.js [hunk 1]
```diff
@@ -0,0 +1 @@
+export const CATEGORIES = "CATEGORIES";
```
Comment (MarcinCholewka) [inline frontend/src/store/actionTypes.js]: try to use semantic names like `GET_ALL_CATEGORIES` or `SAVE_CATEGORIES`

File: frontend/src/store/actions/actions.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    categories
+  };
+};
+
+export const getCategories = () => {
+  return async dispatch => {
+    const categoriesArray = await fetchData("/api/categories");
+    dispatch(saveCategories(categoriesArray));
+  };
+};
```
Comment (MarcinCholewka) [inline frontend/src/store/actions/actions.js]:
```suggestion
export const getCategories = () => {
  return async dispatch => {
    const categoriesArray = await fetchData("/api/categories");
    
    dispatch(saveCategories(categoriesArray));
  };
};
```

File: frontend/src/store/reducers/reducers.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  switch (action.type) {
+    case actionTypes.CATEGORIES:
+      return {
+        ...state,
+        categories: action.categories
+      };
+    default:
+      return {
+        ...state
+      };
```
Comment (MarcinCholewka) [inline frontend/src/store/reducers/reducers.js]: always live empty line before default
```suggestion
      };
      
    default:
      return {
        ...state
      };
```

File: frontend/src/App.js [hunk 4]
```diff
@@ -1,24 +1,28 @@
 import React, { useState, useEffect } from "react";
 import { BrowserRouter as Router, Route } from "react-router-dom";
 import "./App.css";
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.js]: done

File: frontend/src/App.js [hunk 5]
```diff
@@ -1,24 +1,28 @@
...
+import { useDispatch } from "react-redux";
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.js]: done

File: frontend/src/authenticated/AddPost.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    const post = {
+      body: markdown,
+      title: title,
+      categoriesIds: userCategories
+    };
+    const savePost = await postData(
+      "/api/posts",
+      JSON.stringify(post)
+    );
+    if (savePost) history.push("/");
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/authenticated/AddPost.js]: ok

File: frontend/src/authenticated/AddPost.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  };
+
+  const handleInput = input => {
+    setMarkdown(input);
+    if (input.length) setButtonState(false);
+    else setButtonState(true);
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/authenticated/AddPost.js]: done

File: frontend/src/store/actionTypes.js (hunk 1 above)
Comment (LukaszBielsko) [reply frontend/src/store/actionTypes.js]: done

File: frontend/src/App.css [hunk 8]
```diff
@@ -2,32 +2,47 @@
...
   display: flex;
-  flex-direction: column;
-  list-style: none;
+  justify-content: center;
+  margin: 10px auto;
+  width: 80%;
 }
 
 .noShow {
   display: none !important;
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.css]: I'm passing this class to Markdown Editor to hide buttons, that we do not need. 
Cannot do it other way, I'm afraid.

File: frontend/src/authenticated/AddPost.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    );
+
+    if (savePost) {
+      history.push("/");
+    }
+  };
+
+  const handleInput = input => {
+    setMarkdown(input);
+    if (input.length) {
```
Comment (MarcinCholewka) [inline frontend/src/authenticated/AddPost.js]: add empty line before if statement

File: frontend/src/authenticated/EditPost.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  const categoriesOptions = useSelector(state =>
+    convertToSelectOptions(state.categories)
+  );
+  const history = useHistory();
+  const { id } = useParams();
+
+  useEffect(() => {
+    const fetchPost = async () => {
+      const post = await fetchSinglePost("/api/posts/", id);
+      setMarkdown(post.body);
```
Comment (MarcinCholewka) [inline frontend/src/authenticated/EditPost.js]: add empty line before setMarkdown method

File: frontend/src/authenticated/EditPost.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      setTitle(post.title);
+    };
+    fetchPost();
```
Comment (MarcinCholewka) [inline frontend/src/authenticated/EditPost.js]: add empty line before fetchPost method

File: frontend/src/store/reducers/reducers.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import * as actionTypes from "../actionTypes";
+
+const initialState = { categories: [] };
```
Comment (MarcinCholewka) [inline frontend/src/store/reducers/reducers.js]:
```suggestion
const initialState = {
    categories: [],
};
```

File: frontend/src/store/actions/actions.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+export const getCategories = () => {
+  return async dispatch => {
+    const categoriesArray = await fetchData("/api/categories");
+
+    dispatch(saveCategories(categoriesArray));
+  };
+};
```
Comment (MarcinCholewka) [inline frontend/src/store/actions/actions.js]: A bit confusing, saveCartegories using GET_ALL_CATEGORIES.
Change names in these two methods.
Actions name, should be based on type name.

```suggestion
const getAllCategories = categories => {
  return {
    type: actionTypes.GET_ALL_CATEGORIES,
    categories
  };
};

export const saveAllCategories = () => {
  return async dispatch => {
    const categoriesArray = await fetchData("/api/categories");

    dispatch(getAllCategories(categoriesArray));
  };
};
```

File: frontend/src/components/CopyURL.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    let currentURL = getCurrentURL();
+
+    if (postId) {
+      currentURL = currentURL + `posts/${postId}`;
+    }
+
+    navigator.clipboard.writeText(currentURL);
+  };
+
+  return <button onClick={copyURL}>get Url</button>;
```
Comment (MarcinCholewka) [inline frontend/src/components/CopyURL.js]: suggestion => maybe change that text to smth like
```suggestion
  return <button onClick={copyURL}>Copy post url</button>;
```

**Pull request merged**